util/src/main/java/org/hyperledger/besu/util/era1/Era1Reader.java

```diff
@@ -0,0 +1,130 @@
+  private static final int SLOT_INDEX_COUNT_LENGTH = 8;
+
+  private final SnappyFactory snappyFactory;
+
+  /**
+   * Creates a new Era1Reader with the supplied SnappyFactory
+   *
+   * @param snappyFactory A factory to provide objects for snappy decompression
+   */
+  public Era1Reader(final SnappyFactory snappyFactory) {
+    this.snappyFactory = snappyFactory;
+  }
+
+  /**
+   * Reads the entire supplied InputStream, calling appropriate methods on the supplied
+   * Era1ReaderListener as different parts of the file are read
+   *
+   * @param inputStream The InputStream
+   * @param listener the Era1ReaderListener
+   * @throws IOException If there are any problems reading from the InputStream, or creating and
+   *     using other streams, such as a SnappyFramedInputStream
+   */
+  public void read(final InputStream inputStream, final Era1ReaderListener listener)
+      throws IOException {
+    int slot = 0;
+    while (inputStream.available() > 0) {
+      Era1Type type = Era1Type.getForTypeCode(inputStream.readNBytes(TYPE_LENGTH));
+      int length = (int) convertLittleEndianBytesToLong(inputStream.readNBytes(LENGTH_LENGTH));
+      switch (type) {
+        case VERSION -> {
+          // do nothing
+        }
+        case EMPTY, ACCUMULATOR, TOTAL_DIFFICULTY -> {
+          // skip the bytes that were indicated to be empty
+          // TODO read ACCUMULATOR and TOTAL_DIFFICULTY properly?
+          inputStream.skipNBytes(length);
+        }
+        case COMPRESSED_EXECUTION_BLOCK_HEADER -> {
+          byte[] compressedExecutionBlockHeader = inputStream.readNBytes(length);
+          try (SnappyFramedInputStream decompressionStream =
+              snappyFactory.createFramedInputStream(compressedExecutionBlockHeader)) {
+            listener.handleExecutionBlockHeader(
+                new Era1ExecutionBlockHeader(decompressionStream.readAllBytes(), slot));
+          }
+        }
+        case COMPRESSED_EXECUTION_BLOCK_BODY -> {
+          byte[] compressedExecutionBlock = inputStream.readNBytes(length);
+          try (SnappyFramedInputStream decompressionStream =
+              snappyFactory.createFramedInputStream(compressedExecutionBlock)) {
+            listener.handleExecutionBlockBody(
+                new Era1ExecutionBlockBody(decompressionStream.readAllBytes(), slot));
+          }
+        }
+        case COMPRESSED_EXECUTION_BLOCK_RECEIPTS -> {
+          byte[] compressedReceipts = inputStream.readNBytes(length);
+          try (SnappyFramedInputStream decompressionStream =
+              snappyFactory.createFramedInputStream(compressedReceipts)) {
+            listener.handleExecutionBlockReceipts(
+                new Era1ExecutionBlockReceipts(decompressionStream.readAllBytes(), slot++));
+          }
+        }
+        case BLOCK_INDEX -> {
+          ByteArrayInputStream slotIndexInputStream =
+              new ByteArrayInputStream(inputStream.readNBytes(length));
+          long startingSlot =
```

Comment:
rename to blockIndex and same above for slotIndexInputStream as we are only dealing with block indexes and not slots